src/charm.py

```diff
@@ -24,53 +21,193 @@
 
 VALID_LOG_LEVELS = ["info", "debug", "warning", "error", "critical"]
 USG_PACKAGE = "usg"
+AUDIT_HTML_RESULTS_PATH = "/tmp/audit.results.html"
+AUDIT_XML_RESULTS_PATH = "/tmp/audit.results.xml"
 
 
 class CharmCisHardeningCharm(ops.CharmBase):
-    """Charm the service."""
+    """A charm for implementing CIS hardening standards on units."""
+
+    _stored = ops.framework.StoredState()
 
     def __init__(self, framework: ops.Framework):
         super().__init__(framework)
-        # framework.observe(self.on.config_changed, self._on_config_changed)
+        self._stored.set_default(hardening_status=False)
+
+        framework.observe(self.on.config_changed, self._on_config_changed)
         framework.observe(self.on.execute_cis_action, self._cis_harden_action)
         framework.observe(self.on.install, self._on_install)
         framework.observe(self.on.execute_audit_action, self._on_audit_action)
+        framework.observe(self.on.start, self._on_start)
 
     def _on_install(self, event):
-        self.install_usg()
-        self.unit.status = ops.ActiveStatus("Ready to execute CIS hardening.")
-        if self.model.config["auto-harden"]:
-            self.cis_harden()
+        try:
+            self.unit.status = ops.MaintenanceStatus("Installing dependencies...")
+            self.install_usg()
+            self.unit.status = ops.ActiveStatus("Ready for CIS hardening. Run 'execute-cis' action")
+
+            if self.model.config["auto-harden"]:
+                self.unit.status = ops.MaintenanceStatus("Auto-hardening enabled, starting hardening...")
+                self.cis_harden()
+        except Exception as e:
+            logger.error(f"Installation failed: {str(e)}")
+            self.unit.status = ops.BlockedStatus(f"Install failed: {str(e)}")
+
+    def check_state(self):
+        if self._stored.hardening_status:
+            self.unit.status = ops.ActiveStatus("Unit is hardened. Use 'execute-audit' action to check compliance")
+            return
+        if not self.is_configuration_set("tailoring-file"):
+            logger.error("Tailoring-file is not set")
+            self.unit.status = ops.BlockedStatus("Cannot run hardening. Please configure a tailoring-file")
+        else:
+            self.unit.status = ops.ActiveStatus("Ready for CIS hardening. Run 'execute-cis' action")
+
+    def _on_start(self, event):
+        # Workaround needed https://chat.canonical.com/canonical/pl/rr9su5ceh3r98r5jbiuu6989wr
```

Comment:
lets remove chat links - do we have a launchpad bug for people who are external?

Reply:
Removed. No there is no launchpad bug yet for it